debug_module.tex

```diff
@@ -244,6 +244,13 @@ \section{Program Buffer} \label{programbuffer}
 instruction, or a compressed instruction in the lower 16 bits accompanied by a
 compressed {\tt nop} in the upper 16 bits.
 
+\begin{commentary}
+    The slightly inconsistent behavior with a Program Buffer of size 1 is to
+    accommodate hardware designs that prefer to stuff instructions directly
+    into the pipeline when halted, instead of having the Program Buffer exist
+    in the address space somewhere.
```

Comment:
so is this specifically saying it MUST be a NOP?

So is it not a legal implementation to build a program buffer that is only 4-bytes of actual memory based RAM? I guess that wasn't even obvious to me when writing the spec, since can't one do quite a bit (e.g. read a CSR) with a 32-bit program buffer?

I guess maybe the confusion is what does it mean "the program buffer may hold only a single instruction". I read that as "the program buffer is maybe only capable of holding a single instruction", but in your commentary I'm realizing maybe you meant "the program buffer shall hold only a single instruction". What does it mean then if the debugger tries to write 2 compressed instructions? Is Debug Module supposed to quash the upper one? 

What I'm getting at is who are these restrictions imposed on, the HW implementation or the debugger?

Reply:
You can build a program buffer that is 4 bytes of RAM, but the debugger can not write 2 compressed instructions into it.

The debugger must not write 2 compressed instructions. If it does it's out of spec and behavior is undefined. (I'd expect both to be executed, or only the one in the lower 16 bits to be executed, but the debugger mustn't do that.)